src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -625,8 +643,40 @@ private static bool CanBindToTheseConstructorParameters(ParameterInfo[] construc
                 return;
             }
 
-            MethodInfo tryGetValue = dictionaryType.GetMethod("TryGetValue", DeclaredOnlyLookup)!;
-            PropertyInfo indexerProperty = dictionaryType.GetProperty("Item", DeclaredOnlyLookup)!;
+            Func<object, object?> getValue;
+            Action<object, object?> setValue;
+            // Dictionary<TKey, TValue> implements non-generic IDictionary interface. Calling
+            // IDictionary methods is faster than calling IDictionary<T> methods via Reflection.
+            if (dictionary is IDictionary d)
+            {
+                getValue = k =>
+                {
+                    // Unlike IDictionary<TKey, TValue>, non-generic IDictionary should return null
+                    // if key is missing from the dictionary, but we also catch KeyNotFoundException
+                    // as a fail-safe if the implementation is incorrect.
+                    try
+                    {
+                        return d[k];
+                    }
+                    catch (KeyNotFoundException)
+                    {
+                        return null;
+                    }
+                };
+                setValue = (k, v) => d[k] = v;
+            }
+            else
+            {
+                MethodInfo tryGetValue = dictionaryType.GetMethod("TryGetValue", DeclaredOnlyLookup)!;
+                PropertyInfo indexerProperty = dictionaryType.GetProperty("Item", DeclaredOnlyLookup)!;
+
+                getValue = k =>
+                {
+                    object?[] tryGetValueArgs = { k, null };
+                    return (bool)tryGetValue.Invoke(dictionary, tryGetValueArgs)! ? tryGetValueArgs[1] : null;
+                };
+                setValue = (k, v) => indexerProperty.SetValue(dictionary, v, new object[] { k });
```

Comment:
Is it possible to reuse some array (lazy initialized local variable?) to avoid allocation for `new object[] { k }`? Or it is not worth it at all?

Reply:
It should be possible, although I tried to be as close to the previous implementation as I could here. I'll try to make the array reusable and see if it breaks anything.

Reply:
I've looked at code once more and it shouldn't break anything. The setValue is used on one place and it cannot overlap. 
Same should be for `addElement = e => addMethod?.Invoke(collection, new[] { e });`.